node/src/main/scala/coop/rchain/node/instances/ProposerInstance.scala

```diff
@@ -0,0 +1,80 @@
+package coop.rchain.node.instances
+
+import cats.effect.Concurrent
+import cats.effect.concurrent.{Deferred, MVar, Ref, Semaphore}
+import coop.rchain.casper.Casper
+import coop.rchain.casper.blocks.proposer.{BugError, ProposeResult, Proposer}
+import coop.rchain.casper.protocol.BlockMessage
+import coop.rchain.casper.state.instances.ProposerState
+import coop.rchain.shared.Log
+import cats.syntax.all._
+import fs2.Stream
+import fs2.concurrent.Queue
+
+object ProposerInstance {
+  def create[F[_]: Concurrent: Log](
+      proposeRequestsQueue: Queue[F, (Casper[F], Deferred[F, Option[Int]])],
+      proposer: Proposer[F],
+      state: Ref[F, ProposerState[F]]
+  ): Stream[F, (ProposeResult, Option[BlockMessage])] = {
+    // stream of requests to propose
+    val in = proposeRequestsQueue.dequeue
+
+    // max number of concurrent attempts to propose. Actual propose can happen only one at a time, but clients
+    // are free to make propose attempt. In that case proposeID returned will be None.
+    val maxConcurrentAttempts = 100
+
+    val out = Stream
+    // propose permit
+      .eval(for {
+        lock    <- Semaphore[F](1)
+        trigger <- MVar[F].of()
+      } yield (lock, trigger))
```

Comment:
Please check this recent change - I've added this trigger to make sure no propose events triggered are left without following propose attempt.